pkg/services/ngalert/state/cache.go

```diff
@@ -15,28 +15,40 @@ import (
 	ngModels "github.com/grafana/grafana/pkg/services/ngalert/models"
 )
 
+type ruleStates struct {
```

Comment:
Does this need to be a struct with a field or can it just be a `type ruleStates map[string]*State`?

Reply:
You can still write functions on such a type:

```go
type ruleStates map[string]*State

func (rs ruleStates) getOrCreate(key string) *State {
  s, ok := rs[key]
  if !ok { return nil }
  return s
}
```

Then you can simplify a few lines where you have to refer to a `states` field like:
```go
 	ruleStates, ok := c.states[orgID][alertRuleUID]
		return state, nil
	if ok {
		var state *State
		state, ok = ruleStates.states[stateId]
		if ok {
			return state
		}
	}
}
```
back to their simpler form without checking the `ok` value twice.

https://goplay.tools/snippet/Rpf5kfxc0Xb

Reply:
I would like to keep it struct because I will add more fields to it in my other PR